// astro/dist/vite-plugin-fetch/index.js
import MagicString from "magic-string";
function isSSR(options) {
  if (options === void 0) {
    return false;
  }
  if (typeof options === "boolean") {
    return options;
  }
  if (typeof options == "object") {
    return !!options.ssr;
  }
  return false;
}
const SUPPORTED_FILES = /\.(astro|svelte|vue|[cm]?js|jsx|[cm]?ts|tsx)$/;
const IGNORED_MODULES = [/astro\/dist\/runtime\/server/, /\/node-fetch\//];
const DEFINE_FETCH = `import fetch from 'node-fetch';
`;
function pluginFetch() {
  return {
    name: "@astrojs/vite-plugin-fetch",
    enforce: "post",
    async transform(code, id, opts) {
      const ssr = isSSR(opts);
      if (!ssr) {
        return null;
      }
      if (!id.match(SUPPORTED_FILES)) {
        return null;
      }
      if (!code.includes("fetch")) {
        return null;
      }
      for (const ignored of IGNORED_MODULES) {
        if (id.match(ignored)) {
          return null;
        }
      }
      const s = new MagicString(code);
      s.prepend(DEFINE_FETCH);
      const result = s.toString();
      const map = s.generateMap({
        source: id,
        includeContent: true
      });
      return { code: result, map };
    }
  };
}
export {
  pluginFetch as default
};
